Templating  Dockerfiles - JAVA

CooperLink opened the pull request:
This PR is the first in a series of PRs to simplify the process of writing and changing our docker files.  The goal here is to introduce a common base image for the Azure Host runtime bits that can be shared across all of our images.  The method I achieve that through here is by running a script to combine our language specific images with the base image defined in this PR.

Re-using the same base image has two major benefits.  The most pronounced is to reduce the need for scripts like update-version.sh or github commits that require changes to all of our host base images.  With our current configuration, changes to the base image require over 50 file change just to maintain consistency between images.  This is dangerous as there is not an automated method to change all files at once (other than hostversion) and it is easy to miss one or two files while trying to alter all of them.  

Base image re-use hopefully reduces the need for PRs like this : https://github.com/Azure/azure-functions-docker/pull/515/files

By introducing this method, it should be much easier to change our images and maintain consistency.  A lack of consistency is an important barrier to overcome as inconsistent base images invalidate the docker cache leading to unnecessarily longer build times, quicker worker cache exhaustion, and possibly hard to track bugs.

This PR and the forthcoming PRs to apply this everywhere have limited expected impact on cache size but, will hopefully boost developer productivity, traceability and maintainability of this repo. There are numerous PRs in the history of this repo that were required to fix dockerfiles that were accidentally ignored during initial work.  This tool aims to change that. 

Another major benefit of this approach is breaking apart multi-stage builds.  This makes it easier to produce early stages as images which can allow us to cut down on image size by inspecting the runtime-image stage file system and ensuring we are only copying over what is needed. 

### PR information
<!-- You can mark the following checkboxes as [x] to mark them during the PR creation itself. -->
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and CI is passing.

<!-- Thanks for using the checklist -->

Commit:
combine logical units in java

Commit:
Completed Java testing

pragnagopa commented on host/3.0/buster/amd64/base/out/java/java11-appservice.Dockerfile:
```diff
@@ -0,0 +1,73 @@
+
+RUN BUILD_NUMBER=$(echo ${HOST_VERSION} | cut -d'.' -f 3) && \
+    git clone --branch v${HOST_VERSION} https://github.com/Azure/azure-functions-host.git /src/azure-functions-host && \
+    cd /src/azure-functions-host && \
```
no need to build host again here?

ahmelsayed replied:
This is the out directory. Maybe it should be in the root of the repo so it's clearer? i.e:

```
out/host/3.0/amd64/java/java11-appservice.Docker
```

CooperLink replied:
Thank you for the suggestion Ahmed.  I will probably use that path when I am finished implementing this across versions. My plan is to bump it up to all languages v3 -appservice.  Then see what I can do about the other images we have in v3. Then update it for v4. 
I wrote this down and will tag you when I have a PR that incorporates it. 

https://github.com/Azure/azure-functions-docker/issues/538

pragnagopa commented on host/3.0/buster/amd64/base/out/java/java11-appservice.Dockerfile:
```diff
@@ -0,0 +1,73 @@
+    mv /azure-functions-host/workers /workers && mkdir /azure-functions-host/workers && \
+    rm -rf /root/.local /root/.nuget /src
+
+RUN EXTENSION_BUNDLE_VERSION=1.8.1 && \
```
same here - extension bundle is part of host.dockerfile - remove?

TsuyoshiUshio commented on host/3.0/buster/amd64/base/generate-composite.sh:
```diff
@@ -0,0 +1,54 @@
+
+    printf $"\t Collecting global arguments in $1 to globalize them in the final dockerfile.\n"
+    globalize_args $1
+    printf '%s\n' "${gblarg[@]}" > ./temp.txt
```
Nit: You can use >> as append mode. 
However, it is up to you. current code is fine as well.

CooperLink replied:
I **think** that I want to not use append mode here as every individual lang/version will overwrite this file.

ahmelsayed replied:
You can do 

```
echo "${gblarg[@]}" | cat - $base $1 > ./out/${runtime}/${runtimeversion}-appservice.Dockerfile
```

To avoid the need for a tmp file.

CooperLink replied:
Incorporated this in most recent commit. Thank you both

TsuyoshiUshio commented on host/3.0/buster/amd64/base/host.Dockerfile:
```diff
@@ -0,0 +1,39 @@
+    rm -rf /root/.local /root/.nuget /src
+
+RUN EXTENSION_BUNDLE_VERSION=1.8.1 && \
+    EXTENSION_BUNDLE_FILENAME=Microsoft.Azure.Functions.ExtensionBundle.1.8.1_linux-x64.zip && \
```
This line should be replaced by the variables. I mean 1.8.1 part.

CooperLink replied:
I saw this and agree. Will work on extracting all things possible to env vars in the near future.

CooperLink replied:
Including this in a future work item and closing conversation
https://github.com/Azure/azure-functions-docker/issues/539

TsuyoshiUshio reviewed (commented):
Except for 1 comment, It looks great! Thank you for the contribution.  One thing I want to discuss is, if we should include the generated Dockerfiles or not. We saw the both approaches. If we don't include it, PR looks simple and don't see many changes. If we include it as is, it is good since we can see the generated image and which change is introduced to each Dockerfiles. however, people could confuse if they change it or not. However, if you automate it, it might doesn't matter. :)

ahmelsayed commented on host/3.0/buster/amd64/base/temp.txt:
```diff
@@ -0,0 +1 @@
+ARG JAVA_VERSION=8u252
```
remove

CooperLink replied:
Going to .gitignore all the /out/files and can update the script to delete this file at the end.

CooperLink replied:
Saw your comment with a command that could replace the need for this file. Incorporating.

CooperLink replied:
Introduced the .gitignore and removed need for this file

ahmelsayed commented on host/3.0/buster/amd64/java/java11/java11-composite.Dockerfile:
```diff
@@ -0,0 +1,33 @@
+## Using this method with ../ requires the docker build be run from the /out/ folder.
+## Alternative solution would be to copy the main files for sshd and start.sh to each lang folder
+COPY ../shared/sshd_config /etc/ssh/
+COPY ../shared/start.sh /azure-functions-host/
```
`..` doesn't really work in `Dockerfile` as you mentioned in the comment the docker build command needs to run from the `./out` dir. `../shared` will resolve to `/shared` in the current directory

CooperLink replied:
Agreed.  Was just trying to think of a way to fix this.  I see two solutions. 
1: Move all output dockerfiles to the same base /generated/ dir
         Would allow us to just use the two config files in all the dirs instead of copying over lots of them
2: Copy configs to the language folders from a main copy in the /base/ folder
        Copying over this file into all the language folders would also simplify this problem. But, we would create lots of identical files.  Copying them all from the main copy in the /base/ folder would allow us to change all the configs at once similar to the base image.

CooperLink replied:
Chose to copy all configs over to the output folder during ./generate-composite.sh. Will push latest updates with these changes.

ahmelsayed commented on host/3.0/buster/amd64/java/java11/java11-composite.Dockerfile:
```diff
@@ -0,0 +1,33 @@
+COPY ../shared/start.sh /azure-functions-host/
+COPY --from=runtime-image [ "/FuncExtensionBundles", "/FuncExtensionBundles" ]
+
+ENV JAVA_HOME /usr/lib/jvm/zre-11-azure-amd64
```
nit: move with the list of `ENV` above

pragnagopa commented:
As discussed offline - Please checkin the produced docker files into a test directory and add unit tests

Commit:
Template files, test dir, gitignore, remove temp

Commit:
Included sharedconfig to simplify build context

pragnagopa commented on host/3.0/buster/amd64/node/node10/node10-composite.template:
```diff
@@ -0,0 +1,42 @@
+    libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 \
+    libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget
+
+ENV AzureWebJobsScriptRoot=/home/site/wwwroot \
```
except installing specific version of node, rest of the commands are common to all node images. This would be true for python as well -- all the common dependencies in python images except for the python version 

Does it make sense to have template per language as well?

CooperLink replied:
I can definitely do that. In the next PR that I update here I will add a function to do that based on flags for languages that can benefit from it.  For now I will keep it as just to release changes iteratively.  Will probably start by adding this for all v3 images. Then seeing if I can consolidate for node and python/ introduce vars that can be set to create each version.  Imo out of scope for this specific work. Focusing on base image extraction first. :)

pragnagopa replied:
makes sense. Do track this separately - open issue / WI

pragnagopa commented on host/3.0/buster/amd64/node/node10/node10-composite.template:
```diff
@@ -0,0 +1,42 @@
+    apt-get install -y nodejs
+
+# Chrome headless dependencies
+# https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#chrome-headless-doesnt-launch-on-unix
```
Can installing `Chrome headless dependencies` move to host.dockerfile?

CooperLink replied:
Unsure, only because as far as I understand node and python are the only ones that have it.  I will investigate what I can do to consolidate this. 
https://github.com/Azure/azure-functions-docker/pull/417

You also made a comment in this PR to see if these could be moved to the main dockerfile.

pragnagopa replied:
Thanks for linking earlier PR - makes sense to move this to common file as it does not seem to language specific - we can track this separately as well. 

We should move this to host.dockerfile as long as it does not increase image size drastically as Node and Python images need it anyways

Commit:
Updated Java Build Pipeline

Commit:
Updated execution permissions

Commit:
Updated host script to run anywhere

Commit:
Remove node, update dotnetsdk, testing java images

Commit:
Updated flag with release mode. Removed test

Commit:
Update release to generate images

CooperLink commented:
I had a good discussion with Bala about traceability of the dockerfiles. With the solution proposed here, we can still view git history of the files and understand changes.  But, changes to the .Dockerfiles that actually build the images that we use in production (java8-appservice.Dockerfile) would be harder to visualize. 
To fix this problem, I added a switch -r that generates images in the release folder on a branch.  I also updated the release drafter to generate those images.  So when we cut a release it will include a snapshot of the images that that version is using in production. This will also help out with a task I am doing soon to rebuild the images every two weeks.

pragnagopa commented:
Adding notes from offline conversation:
- Include generated docker files as " Build Artifacts"
- Upload Build Artifacts to github release page as part of release pipeline

Commit:
Generate release artifacts

Commit:
Manually remove /out/ folders

Commit:
Remove now un-used images

CooperLink commented:
I will be squashing and merging this branch after approval :) Just rebased.

pragnagopa reviewed (approved):
LGTM! @TsuyoshiUshio / @balag0  - please take a final look as well! @CooperLink  - please open follow up github issues for unresolved comments

balag0 commented:
Looks great. Greatly simplifies updating host versions. One minor thing for later - if there is any guidance for new dockerfile authors / making more involved changes in existing dockerfiles, please document them somewhere.

TsuyoshiUshio commented on host/3.0/buster/amd64/base/host.Dockerfile:
```diff
@@ -23,7 +23,7 @@ RUN EXTENSION_BUNDLE_VERSION=1.8.1 && \
     unzip /$EXTENSION_BUNDLE_FILENAME -d /FuncExtensionBundles/Microsoft.Azure.Functions.ExtensionBundle/$EXTENSION_BUNDLE_VERSION && \
     rm -f /$EXTENSION_BUNDLE_FILENAME && \
-    EXTENSION_BUNDLE_VERSION_V2=2.8.3 && \
+    EXTENSION_BUNDLE_VERSION_V2=2.6.1 && \
```
I'm curious about this downgrade. Could you share the context?

pragnagopa replied:
good catch @TsuyoshiUshio  - this is a recent update . PR is needs to be udpated.

TsuyoshiUshio reviewed (approved):
Great Job! Thank you for the all of effort and clean bash code.

pragnagopa reviewed (dismissed):
@CooperLink - Lets fix the extension bundle version and merge the PR

Commit:
Updated ext and host ver
